Horoscope panel

domenc opened the pull request:
Issue #209

Implement Horoscope panel (slovenian and english). Slovenina works only if translated horoscope mark equaly as in slovenskenovice.si

domenc committed:
Horoscope panel (slovenian and english)

domenc commented:
Waiting cor comment :)

mitar commented:
I don't know where to start. ;-)

Definitely we do not want to create external request on every rendering of a panel for every request to home page for every user! You should define background Celery task which fetches data and stores it into the database. And then you render that.

Then, for data sources we can use only those who allow this. Please find information that we are allowed to re-publish it and under what terms. For Slovenske novice we could probably ask for permission.

Then, split conversion between date and sign, and translation of this sign. Translation is something else, conversion is something else.

mitar commented:
OK, this for now. Then there are some comments on code structure.

mitar commented:
Are there any open/free/creative commons for signs to use?

mitar commented:
I sent an e-mail to Slovenske novice if we could use their horoscope.

domenc commented:
How to create celery task? And how to insert and read from databse :)
Now, I split date and sign for translation and for requesting.

mitar commented:
You define it in file named `tasks.py`, configure that in settings. For database read MongoEngine documentation. See how we are doing elsewhere in PiplMesh.

mitar commented:
`models.py` inside your panel app.

domenc committed:
Add horoscope model for database
Edit css, html file - display horoscope only if user is login and have set his birthdate
Add horoscope task: every day at 6 am get horoscope for all available languages for all marks (id horoscope in database doesnt exits (by languages and mark field then get inser it, else update fields))

Edit API for horoscope

domenc commented:
Add horoscope model for database (fields: language, mark, description, source).
Edit css, html file - display horoscope only if user is login and have set his birthdate.
Add horoscope task: every day at 6 am get horoscope for all available languages for all marks (if horoscope in database doesn't exits (by language and mark field then get inser it, else update fields with new data))

Edit API for horoscope

mitar commented on piplmesh/frontend/tasks.py:
```diff
@@ -1,4 +1,5 @@
 import datetime
+import itertools
```
You can put those two into one line.

mitar commented on piplmesh/frontend/tasks.py:
```diff
@@ -10,6 +11,9 @@
 from piplmesh.frontend import views
 
+from piplmesh.panels.horoscope import horoscope
+from piplmesh.panels.horoscope.models import Horoscope
```
Import only modules!

mitar commented on piplmesh/frontend/tasks.py:
```diff
@@ -59,3 +63,19 @@ def check_online_users():
             )
+
+@task.task
+def update_horoscope():
```
NOOO. Move this to `tasks.py` in your panel app, not to frontend. So along your `models.py`, add `tasks.py`.

mitar commented on piplmesh/panels/horoscope/horoscope.py:
```diff
@@ -0,0 +1,194 @@
+    """
+    Function return horoscope mark number.
+    Returnd numbers:
+        0 - Aries
```
I would prefer that you return lower case constant strings.

domenc replied:
Its easier to select name for displaying :)

.. I see next comment :)

mitar commented on piplmesh/panels/horoscope/horoscope.py:
```diff
@@ -0,0 +1,194 @@
+    Return translated horoscope marks.
+    """
+
+    horoscope_marks = [_("Aries"), _("Pisces"), _("Taurus"), _("gemini"), _("Cancer"), _("Leo"), _("Virgo"), _("Libra"), _("Scorpio"), _("Sagittarius"), _("Capricorn"), _("Aquarius")]
```
And then you can have a dict here, mapping from those constant strings to translated strings.

This can go to a top-level module constant (named with ALL_CAPS).

domenc replied:
What do you think with this: This can go to a top-level module constant (named with ALL_CAPS).

mitar replied:
See how we defined choices [here](https://github.com/wlanslovenija/PiplMesh/blob/master/piplmesh/account/fields.py).

domenc replied:
Is this ok? 

language = mongoengine.StringField(choices=settings.LANGUAGES)

domenc replied:
AA, i dont know what you think about with ALL_CAPS :) - caps lock

domenc replied:
But if i have a dict how can get right request_string? Now i have with number of array..

mitar replied:
Just instead of numbers, use some common string constants. For example, lower case English names. And then use that as a key.

mitar commented on piplmesh/panels/horoscope/models.py:
```diff
@@ -0,0 +1,12 @@
+from django.db import models
+
+import mongoengine
+from mongoengine import *
```
Import only modules.

mitar commented on piplmesh/panels/horoscope/models.py:
```diff
@@ -0,0 +1,12 @@
+from __future__ import absolute_import
```
Use this if you need it. You do relative imports anywhere?

mitar commented on piplmesh/panels/horoscope/models.py:
```diff
@@ -0,0 +1,12 @@
+
+class Horoscope(Document):
+    mark = mongoengine.IntField()
+    language = mongoengine.StringField()
```
You can limit possible values of language with `choices`. You can limit to languages available.

mitar commented on piplmesh/panels/horoscope/horoscope.py:
```diff
@@ -0,0 +1,194 @@
+    horoscope_marks = [_("Aries"), _("Pisces"), _("Taurus"), _("gemini"), _("Cancer"), _("Leo"), _("Virgo"), _("Libra"), _("Scorpio"), _("Sagittarius"), _("Capricorn"), _("Aquarius")]
+    return horoscope_marks[mark]
+
+def get_horoscope_available_languages():
```
This is not the best way. Better would be to define a simple base class with few methods. And each of them have methods to retrieve data, and also to tell which language is for. Then you just have a list of those classes. And then a function which returns list of supported languages based on those registered classes.

In this way, new language can be made only be creating a new subclass of the base class and adding to the list. It is not necessary to go all around the code and search where it should also be changed.

domenc replied:
Is any example in piplmesh projects?

mitar replied:
Panels themselves are such an example. ;-) But there is a bit more complicated, because they allow registering from different apps. You do not need this (you do not need something like `panels_pool` we have). But for structure of base class and subclasses, yes, something like that.

domenc replied:
Doens't know how to create this with subclass

mitar replied:
Start with creating a class and defining some methods in, those you have now all around and are connected with implementation for one given language.

domenc replied:
If i undestand right:

1.)
class OneLanguage:
 def getDescription:
 def getData:

??

mitar replied:
```
class HoroscopeBase(object):
    ...


class SlovenianHoroscope(HoroscopeBase):
    def get_language_code(self):
        ...

    def fetch_data(self):
        ....
```

mitar commented on piplmesh/panels/horoscope/horoscope.py:
```diff
@@ -0,0 +1,194 @@
+    If we create a new function for get horoscope in a new languages, we must add language in below list.
+    """
+
+    available_languages = ['en', 'sl']
```
Use tuples for such constant lists.

mitar commented on piplmesh/panels/horoscope/horoscope.py:
```diff
@@ -0,0 +1,194 @@
+        11 - Aquarius
+    """
+
+    if month == 3:
```
Are you sure there does not exist any existing Python library for this?

mitar commented on piplmesh/settings.py:
```diff
@@ -14,6 +14,8 @@
 import djcelery
 djcelery.setup_loader()
 
+from celery.task.schedules import crontab
```
Oh, I didn't know about this one. Can you port also other periodic tasks to use this function? And remove timedelta.

mitar commented on piplmesh/settings.py:
```diff
@@ -209,6 +211,7 @@
 }
 
 CHECK_ONLINE_USERS_INTERVAL = 10
```
And also add units in a comment at the end, like:

```
CHECK_ONLINE_USERS_INTERVAL = 10 # seconds
```

mitar commented on piplmesh/panels/horoscope/panel.py:
```diff
@@ -0,0 +1,20 @@
+        (horoscope_description, horoscope_mark, horoscope_source) = get_daily_horoscope(context['user'])
+
+        context.update({
+            'header': _("Today's horoscope"),
```
Hm, somewhere should also be displayed the date of the horoscope. For people to really know for which "today" this is. So this header is good, but somewhere in the body of panel should be a more precise date.

mitar replied:
(Based on date of source, not when we fetched.)

domenc replied:
no, its difficult for parsing from web :)

mitar replied:
But we need it. Otherwise information is useless. We could predict future wrong! For the wrong day!

mitar commented on piplmesh/panels/horoscope/horoscope.py:
```diff
@@ -0,0 +1,194 @@
+    Daily english horoscope from http://findyourfate.com
+    """
+
+    request_mark = ['Aries', 'Pisces', 'Taurus', 'Gemini', 'Cancer', 'Leo', 'Virgo', 'Libra', 'Scorpio', 'Sagittarius', 'Capricorn', 'Aquarius']
```
Use dict.

mitar commented on piplmesh/panels/horoscope/horoscope.py:
```diff
@@ -0,0 +1,194 @@
+    return get_daily_english_horoscope(mark)
+
+def get_daily_english_horoscope(mark):
+    """
```
So move this to a class. And then `request_mark` could be a class attribute.

mitar commented on piplmesh/panels/horoscope/horoscope.py:
```diff
@@ -0,0 +1,194 @@
+    If we create a new function for get horoscope in a new languages, we must add a if statmenet below.
+    """
+
+    if lang == 'sl':
```
No, we should not hardcode this like this. No language should be special. If we do not have language, then we don't display anything. (Or display request for help, that somebody gives us source in that language.)

domenc replied:
How to remove panels if horoscope for languages doesn't exits? I try panels.panels_pool.unregister(HoroscopePanel) but doesnt work under get_context function - where try and remove it?

mitar replied:
Why removing it? Just display some other content in the template.

mitar commented on piplmesh/frontend/tasks.py:
```diff
@@ -59,3 +63,19 @@ def check_online_users():
+            (desc, mark_number, src) = horoscope.celery_horoscope_update(i_mark, i_language)
+
+            try:
+                desc = desc.encode("utf-8")
```
It is strange that you should be doing this here. This should be done in the function which fetches the content.

Also, Django has already some functions to handle such things. Try searching in `django.utils`.

domenc replied:
Do't now how to replace 

try:
                desc = desc.encode("utf-8")
            except UnicodeError:
                desc = unicode(desc, "utf-8")

Always get errors

mitar replied:
What exactly are you trying to solve here. Which errors are you getting? What is input? What you want as output?

domenc replied:
Found it.. desc=u'%s'%unicode(smart_unicode(desc)) and is working :)

mitar replied:
And just `smart_unicode(desc)` is not enough? ;-) Or just `unicode(smart_unicode(desc))`?

domenc replied:
:D it is

mitar replied:
Which one?

domenc replied:
desc=smart_unicode(desc)

mitar commented:
OK, enough for now. Good stuff. We have few iterations more to do. But you are moving into good direction. ;-)

domenc committed:
Task: crontab has no support for seconds so I let it deltatime
Html: display date and if language horoscope doesnt exits return string with error
Horoscope: implemnet class with subclasses

TODO: see in models.py

domenc commented:
Task: crontab has no support for seconds so I let it with deltatime 
Html: display date and if language horoscope doesnt exits return string with error
Horoscope: implemnet class with subclasses and dict

TODO: see in models.py (getting error)

mitar commented on piplmesh/settings.py:
```diff
@@ -235,6 +238,11 @@
         'schedule': datetime.timedelta(seconds=CHECK_ONLINE_USERS_INTERVAL),
         'args': (),
     },
+    'update_horoscope': {
```
Add your module to `CELERY_IMPORTS`.

mitar commented on piplmesh/panels/horoscope/models.py:
```diff
@@ -0,0 +1,13 @@
+from piplmesh import settings
+
+class Horoscope(Document):
+    sign = mongoengine.StringField() #TODO: add choices=piplmesh.panels.horoscope.horoscope.HOROSCOPE_SIGN - return error
```
Of course, choices should be tuple of tuples, not dict. You have to transform it.

domenc replied:
Ok, but get error: 

from piplmesh.panels.horoscope import horoscope
ImportError: cannot import name horoscope

mitar replied:
You probably have cyclic imports. ;-)

domenc replied:
So it can not set a choices??

mitar replied:
You have to break this. In general, you can move common things somewhere else. But in your case you can maybe have a list of signs in models as tuple of tuples. ;-) And then use that for creating a dict in `horoscope.py`.

mitar commented on piplmesh/panels/horoscope/models.py:
```diff
@@ -0,0 +1,13 @@
+import mongoengine
+from mongoengine import Document
+
+from piplmesh import settings
```
No! Check how settings should be imported.

mitar commented on piplmesh/panels/horoscope/horoscope.py:
```diff
@@ -0,0 +1,267 @@
+from xml.dom.minidom import parseString
+
+from django.db import models
+from mongoengine import *
```
Import only modules!

mitar replied:
Please group imports as we are grouping them elsewhere (by library, first build-in, then django, then others, then piplmesh).

mitar commented on piplmesh/panels/horoscope/horoscope.py:
```diff
@@ -0,0 +1,267 @@
+
+def get_horoscope_sign(day, month):
+    """
+    Function return horoscope sign key for map.
```
Just: _Based on date, returns horoscope sign key._

mitar commented on piplmesh/panels/horoscope/horoscope.py:
```diff
@@ -0,0 +1,267 @@
+    @classmethod
+    def get_list_of_languages(cls):
+        languages = []
+        for h_object in HoroscopeBase.__subclasses__():
```
Very very interesting. I didn't know that this exist. But this is ugly. ;-)

domenc replied:
Would be better if for get_list_of_languages(cls) get **init** methods in each of class and they add languages avtomatic into one dict in parent class?

mitar replied:
Main idea is that we do not have too many different ways (patterns) of doing things. So currently we have two ways:
- list of classes in settings (like middleware, authentication, ... is done)
- a pool of classes and registration to it (like panels are done)

The last one is useful when you have multiple Python modules/Django apps contributing automatically together. Nothing is done automatically, so that it is clearly visible from the code what is happening. (You see register function in the panels.)

Now, the best thing would be to implement a similar pool for horoscopes. So that some third-party modules could also add themselves. But this is probably overkill here because we will not probably have much of this.

So I am proposing that instead of listing classes in settings, we do this in the module itself, as this is not really something configurable.

Yes, we could do some smart and crazy thing, but this would then leave other developers wondering ... do I have to register this somewhere? Do I have to configure settings. How do I "enable" this thing. How I "disable" if it is just temporary. So ... Do it in a simple manner with a tuple of references to classes.

mitar commented on piplmesh/panels/horoscope/horoscope.py:
```diff
@@ -0,0 +1,267 @@
+
+def get_daily_horoscope(user_object):
+    """
+    Return horoscope sign, description, date and a source depend on user information.
```
We write in third person singular.

domenc replied:
And how can i write this.. My english is not very good

mitar replied:
I think it is quite good.

_Returns horoscope sign, description, date and a source depending on user information._

So _s_ at the end of the verb. ;-)

domenc replied:
:D

mitar commented on piplmesh/panels/horoscope/horoscope.py:
```diff
@@ -0,0 +1,267 @@
+        else:
+            return 'aquarius'
+
+def get_daily_horoscope(user_object):
```
We often name just `user`. We mostly know that it is an object. ;-)

mitar commented on piplmesh/panels/horoscope/horoscope.py:
```diff
@@ -0,0 +1,267 @@
+    Return horoscope sign, description, date and a source depend on user information.
+    """
+
+    user_language = user_object.language
```
Field access does not cost so much that we would really need a variable. ;-)

mitar commented on piplmesh/panels/horoscope/horoscope.py:
```diff
@@ -0,0 +1,267 @@
+            (desc, user_sign, src, date) = HoroscopeBase.get_horoscope(user_language, user_sign)
+            insert_update_one_horoscope(user_language, desc, user_sign, src, date)
+
+            return (desc, HOROSCOPE_SIGNS[user_sign], src, date)
```
OK, this signals  that this is a bit too complex function. ;-)

I would suggest that you all this move to `get_context` method of the panel.

mitar replied:
So you can populate context in one way if horoscope is available, and in another if not.

domenc replied:
Why is not good this? 
You think that if horoscope doesnt in DB then you get error? I think that if doesnt exits in DB you get fetch data from sites..

mitar replied:
Returning four-element tuple is not good. It shows complicated API. You should define good internal APIs. For example, returning an object. Or in this case, this could could simply be moved to `get_context`.

domenc replied:
OK, i return an object :D (with could could simply be moved to get_context you mean that all code move to panels.py?)

mitar replied:
Code of this function could be moved there, yes.

mitar commented on piplmesh/panels/horoscope/horoscope.py:
```diff
@@ -0,0 +1,267 @@
+
+    return ("", _("Error getting horoscope."), "", "")
+
+def update_all_horoscope():
```
You can also move all this to task function.

mitar commented on piplmesh/panels/horoscope/horoscope.py:
```diff
@@ -0,0 +1,267 @@
+from mongoengine import *
+from mongoengine.django import *
+
+from models import Horoscope
```
Import only modules. Import with relative or absolute. Never like this (what if there is some other module `models` somewhere else in the Python path).

mitar commented on piplmesh/panels/horoscope/horoscope.py:
```diff
@@ -0,0 +1,267 @@
+
+            insert_update_one_horoscope(h_lang, h_desc, h_sign, h_src, h_date)
+
+def insert_update_one_horoscope(h_lang, h_desc, h_sign, h_src, h_date):
```
Helper functions for `update_all_horoscope` also move to tasks.

mitar commented on piplmesh/panels/horoscope/horoscope.py:
```diff
@@ -0,0 +1,267 @@
+    """
+
+    @classmethod
+    def get_list_of_avaiable_horoscope(cls):
```
This should go out of the base class. Make functions in this modules. So then you can import module and do:

```
horoscope.get_all_horoscopes()
```

mitar commented on piplmesh/panels/horoscope/horoscope.py:
```diff
@@ -0,0 +1,267 @@
+        return HoroscopeBase.__subclasses__()
+
+    @classmethod
+    def get_list_of_languages(cls):
```
`get_supported_languages()`

mitar commented on piplmesh/panels/horoscope/horoscope.py:
```diff
@@ -0,0 +1,267 @@
+            h_language = h_object().get_language()
+            if language == h_language:
+                return h_object().fetch_data(sign)
+
```
So instead of `__subclasses__` I would manually define a variable:

```
HOROSCOPES = (
    EnglishHoroscope(),
    SlovenianHoroscope(),
)
```

domenc replied:
Harcoded or on **init** dat they inserted to dict?

mitar replied:
Hard-coded as a top-level variable.

mitar replied:
(Ehm, probably at the bottom of the file, after you define them.)

mitar commented on piplmesh/panels/horoscope/horoscope.py:
```diff
@@ -0,0 +1,267 @@
+        horoscope_xml = url_xml_open.read()
+
+        date = parseString(horoscope_xml).getElementsByTagName('title')[1].toxml()
+        date = date.replace('<title>'+request_sign[sign]+' Horoscope for ', '').replace('</title>', '')
```
You probably should learn to use regex. ;-)

domenc committed:
Fix code (missing regex)

domenc commented:
Fix some code and almost all stuf you say it to fix (except regex)

domenc commented:
Have one idea for new issue - horoscope sign images (not for this issue!) :D

mitar commented:
> Have one idea for new issue - horoscope sign images (not for this issue!) :D

Open it!

mitar commented on piplmesh/panels/horoscope/horoscope.py:
```diff
@@ -0,0 +1,211 @@
+from django.db import models
+from django.utils.translation import ugettext_lazy as _
+
+from piplmesh.panels.horoscope.models import Horoscope
```
Import only modules!!!

mitar commented on piplmesh/panels/horoscope/panel.py:
```diff
@@ -0,0 +1,48 @@
+
+from piplmesh import panels
+from piplmesh.panels.horoscope import horoscope
+from piplmesh.panels.horoscope.models import Horoscope
```
Import only modules!

mitar commented on piplmesh/panels/horoscope/models.py:
```diff
@@ -0,0 +1,28 @@
+from django.utils.translation import ugettext_lazy as _
+
+import mongoengine
+from mongoengine import Document
```
Same.

mitar commented on piplmesh/panels/horoscope/horoscope.py:
```diff
@@ -0,0 +1,211 @@
+    """
+
+    @classmethod
+    def get_supported_languages(cls):
```
Move this to the module, make it a module-level function!

mitar replied:
(This is not Java, that everything has to be in classes.)

mitar replied:
Because this is not really a class method, but a static method.

mitar commented on piplmesh/panels/horoscope/horoscope.py:
```diff
@@ -0,0 +1,211 @@
+        return languages
+
+    @classmethod
+    def get_horoscope(cls, language, sign):
```
Same.

mitar commented on piplmesh/panels/horoscope/horoscope.py:
```diff
@@ -0,0 +1,211 @@
+        """
+
+        for object in HOROSCOPES:
+            h_language = object.get_language()
```
`h_language`? What is `h`? Write verbose/full name variables. Or don't. Here it would be much easier to read if you would have:

```
for horoscope in HOROSCOPES:
    if language == horoscope.get_langage():
        ...
```

mitar commented on piplmesh/panels/horoscope/horoscope.py:
```diff
@@ -0,0 +1,211 @@
+        for object in HOROSCOPES:
+            h_language = object.get_language()
+            if language == h_language:
+                return object.fetch_data(sign)
```
If you are fetching data here, then also name this method/function like this. You have `get_horoscope` but in fact you do `fetch_data`.

Name it then `fetch_horoscope_data`.

mitar replied:
But in fact it would be much better if you would have such API (on the module-level):

```
get_horoscope_provider(language):
    """
    Returns an instance of the horoscope provider for requested language, or raises ``KeyError`` if it does not exist.
    """

def get_supported_languages():
    """
    Returns a list of supported languages (their language codes).

    That is, a list of languages provided by defined horoscope providers.
    """ 
```

mitar replied:
And then you could simply call:

```
horoscope.get_horoscope_provider(language).fetch_data()
```

mitar commented on piplmesh/panels/horoscope/horoscope.py:
```diff
@@ -0,0 +1,211 @@
+
+    language = 'en'
+
+    def get_language(self):
```
This method should be moved to HoroscopeBase. That's why we have base class. ;-)

mitar commented on piplmesh/panels/horoscope/horoscope.py:
```diff
@@ -0,0 +1,211 @@
+        return self.language
+
+    def fetch_data(self, sign):
+        request_sign = {
```
Move this to class-level attribute. Don't have long data definition in function. (You have to define it only once, for the class, not that it has to be evaluated/executed on every function call.)

mitar commented on piplmesh/panels/horoscope/horoscope.py:
```diff
@@ -0,0 +1,211 @@
+        }
+
+        source = "http://findyourfate.com"
+        source_xml = "%s/rss/dailyhoroscope-feed.asp?sign=%s" % (source, request_sign[sign])
```
Why are you defining constant variables you are using only once?

mitar replied:
And name variables better. This is not XML. This is URL. ;-)

So:

```
horoscope_url = 'http://findyourfate.com/rss/dailyhoroscope-feed.asp?sign=%s' % request_sign[sign]
horoscope_xml = urlopen(horoscope_url).read()
```

domenc replied:
we need source for displaying it. it's ugly for long url - display only domain.tld.

Nevermind.. I found below how to do it :)

mitar replied:
Read the rest of comments first. ;-)

domenc replied:
OK, but I also need porvider name - not only url:

<  a href="http://findyyy.com/bla/bla/sss.xml">findyourfate.com<   /a  >?

mitar replied:
Read all comments first. I proposed `get_source_name` method for that.

mitar commented on piplmesh/panels/horoscope/models.py:
```diff
@@ -0,0 +1,28 @@
+    sign = mongoengine.StringField(choices=HOROSCOPE_SIGNS)
+    language = mongoengine.StringField(choices=settings.LANGUAGES)
+    description = mongoengine.StringField()
+    source = mongoengine.StringField()
```
Why is this stored in the database? This is a property of the horoscope provider, not of the content of the daily horoscope.

mitar replied:
Define a method in horoscope providers, so that you can call:

```
horoscope.get_horoscope_provider(language).get_source_name()
horoscope.get_horoscope_provider(language).get_source_url()
```

mitar commented on piplmesh/panels/horoscope/horoscope.py:
```diff
@@ -0,0 +1,211 @@
+
+        return (horoscope, sign, source, date)
+
+HOROSCOPES = (
```
Rename this to `HOROSCOPE_PROVIDERS`

mitar commented on piplmesh/panels/horoscope/horoscope.py:
```diff
@@ -0,0 +1,211 @@
+
+    return HOROSCOPES
+
+class HoroscopeBase(object):
```
Rename this to `HoroscopeProviderBase`.

mitar commented on piplmesh/panels/horoscope/panel.py:
```diff
@@ -0,0 +1,48 @@
+    def get_context(self, context):
+        user=context['user']
+
+        horoscope_description = ""
```
Constant strings.

domenc replied:
You mean that must be in up case?

mitar replied:
No, but that it should be `''`. We are using `''` around constant strings. We use `" ... "` only for strings displayed to the user.

mitar commented on piplmesh/panels/horoscope/panel.py:
```diff
@@ -0,0 +1,48 @@
+
+class HoroscopePanel(panels.BasePanel):
+    def get_context(self, context):
+        user=context['user']
```
Spaces around `=`.

mitar commented on piplmesh/panels/horoscope/panel.py:
```diff
@@ -0,0 +1,48 @@
+        horoscope_sign = ""
+        horoscope_date = ""
+
+        if user.birthdate and user.language in horoscope.HoroscopeBase.get_supported_languages():
```
See, why `horoscope.HoroscopeBase.get_supported_languages()`. Why not simply `horoscope.get_supported_languages()`?

mitar commented on piplmesh/panels/horoscope/panel.py:
```diff
@@ -0,0 +1,48 @@
+
+                horoscope_sign=horoscope.HOROSCOPE_SIGNS_DICT[horoscope_sign]
+        else:
+            horoscope_sign = _("Horoscope is unavailable in your language.")
```
Don't reuse variables!

mitar commented on piplmesh/panels/horoscope/panel.py:
```diff
@@ -0,0 +1,48 @@
+                insert_update_one_horoscope(user.language, horoscope_description, horoscope_sign, horoscope_source, horoscope_date)
+
+                horoscope_sign=horoscope.HOROSCOPE_SIGNS_DICT[horoscope_sign]
+        else:
```
You could turn around the if. You can do that when one part is much shorter than the other. So here you can simply set the context in the "else" case and return. And then do the rest if it is available.

mitar commented on piplmesh/panels/horoscope/panel.py:
```diff
@@ -0,0 +1,48 @@
+                horoscope_sign = horoscope.HOROSCOPE_SIGNS_DICT[horoscope_object[0].sign]
+                horoscope_date = horoscope_object[0].date
+            else:
+                (horoscope_description, horoscope_sign, horoscope_source, horoscope_date) = horoscope.HoroscopeBase.get_horoscope(user.language, user_sign)
```
NO! Don't fetch anything during user request processing. This should be done in the background. What if this takes minutes because third-party server is slow? What if it fails?

mitar replied:
Simply set context in a way that template will display an error. So you have three states:
- we have daily horoscope for today in a given language for a given sign
- we do have a provider, but no data (it is too old, or none)
- we do not have a provider

mitar commented:
You have some comments. Please fix that not only on lines I commented on, but everywhere in your code, if applicable.

domenc committed:
Some fix.

domenc commented:
Fix almost all, but:
- doesnt using horoscope.get_horoscope_provider(language).fetch_data() in tasks, because its double for statmenet (wasting time)
  - all horosope must have 3 variables (language, source_name, source_name_url) and fetch_data methos

How to test in template if user is login (with own usernmae and not with random one)  or not???

mitar commented:
> How to test in template if user is login (with own usernmae and not with random one) or not???

`user.is_authenticated()` returns this. Why do you need it?

mitar commented on piplmesh/panels/horoscope/horoscope.py:
```diff
@@ -0,0 +1,230 @@
+from collections import defaultdict 
```
Import only modules!

mitar commented on piplmesh/panels/horoscope/horoscope.py:
```diff
@@ -0,0 +1,230 @@
+    Returns an instance of the horoscope provider for requested language, or raises ``KeyError`` if it does not exist.
+    """
+
+    if language not in get_supported_languages():
```
No need to do it like this. You just go over all horoscope providers and if you get out of the loop (you have not found any), you just raise the exception.

mitar commented on piplmesh/panels/horoscope/horoscope.py:
```diff
@@ -0,0 +1,230 @@
+    if language not in get_supported_languages():
+        raise KeyError("Unsupported language: '%s'" % language)
+    else:
+        for horoscope_instance in get_all_horoscopes_providers():
```
Why `horoscope_instance`? And not just simply `horoscope`?

mitar commented on piplmesh/panels/horoscope/horoscope.py:
```diff
@@ -0,0 +1,230 @@
+    languages = []
+    for object in get_all_horoscopes_providers():
+        language = object.get_language()
+        if language not in languages:
```
Good idea to check, but make this an assertion.

domenc replied:
I dont know what do you think about this.. assertion??

mitar replied:
Google "Python assertion". ;-)

mitar commented on piplmesh/panels/horoscope/horoscope.py:
```diff
@@ -0,0 +1,230 @@
+            languages.append(language)
+    return languages
+
+def fetch_horoscope_data(language, sign):
```
No, this is ugly! Use the code as I have proposed. The code is much more readable.

mitar commented on piplmesh/panels/horoscope/horoscope.py:
```diff
@@ -0,0 +1,230 @@
+    """
+    Parent class for defined horoscope methods for getting horoscope in your language.
+    """
+
```
Define here default attributes:

```
language = None
source_name = None
source_url = None
```

mitar commented on piplmesh/panels/horoscope/horoscope.py:
```diff
@@ -0,0 +1,230 @@
+
+    language = 'en'
+    source_name = 'Find your fate'
+    source_url = 'http://findyourfate.com/rss/horoscope-feed.asp'
```
No, `source_url` shoud point to the website, not feed URL. So people will click on this link to see who provides us. Not machines.

mitar commented on piplmesh/panels/horoscope/horoscope.py:
```diff
@@ -0,0 +1,230 @@
+    source_url = 'http://findyourfate.com/rss/horoscope-feed.asp'
+
+    request_sign = {
+            'aries' : 'Aries',
```
Indent one less.

mitar commented on piplmesh/panels/horoscope/horoscope.py:
```diff
@@ -0,0 +1,230 @@
+
+    language = 'sl'
+    source_name = 'Slovenske novice'
+    source_url = 'http://www.slovenskenovice.si/lifestyle/astro/'
```
Same.

mitar commented on piplmesh/panels/horoscope/models.py:
```diff
@@ -0,0 +1,26 @@
+    sign = mongoengine.StringField(choices=HOROSCOPE_SIGNS)
+    language = mongoengine.StringField(choices=settings.LANGUAGES)
+    description = mongoengine.StringField()
+    date = mongoengine.StringField()
```
`StringField`? This should be `DateField`.

domenc replied:
Mybe have a problem from string to data convesation (slovenian)..

mitar replied:
Yes, convert. ;-) You should always try to normalize data before storing. And then you output in the way you want it.

domenc replied:
But i excpet that i have probelm from "Oktober" -> "October" or python this know??

mitar replied:
Of course will not know. You have to tell him. ;-)

domenc replied:
Then i create dict?

mitar replied:
You can create a parsing function. How you will implement this ... maybe there is something already on the Internet? Maybe Django has something. Or Python. Check internationalization and parsing of dates.

domenc replied:
Can this be a new issue, because on windows OS is this differently then on linux and mac

mitar replied:
No. This should should be OS independent. You do not need or even do not want to use OS-dependent or OS-language dependent things.

domenc replied:
But, then i can't used locale.setlocale() for parsing slovenian date: datetime.date.strptime(slo, "%d. %B" )...

mitar replied:
Something like this does not work?

```
from django.utils import translation
current_language = translation.get_language()
try:
    translation.activate('sl')
    date = datetime.date.strptime(slo, "%d. %B")
finally:
    translation.activate(current_language)
```

mitar replied:
(See the code on GitHub, I edited it.)

domenc replied:
doesn't work for me..

time data '16. junij' does not match format '%d. %B'

domenc replied:
Can i simply create dict with slovenian and english months name??

mitar replied:
Read here:

https://docs.djangoproject.com/en/dev/topics/i18n/formatting/#locale-aware-input-in-forms
https://code.djangoproject.com/ticket/13339#comment:11

So yes, it seems this will not work. You will have to parse it manually. Yes, you can internally use dict for mappings.

mitar commented on piplmesh/panels/horoscope/panel.py:
```diff
@@ -0,0 +1,50 @@
+from django.utils.translation import ugettext_lazy as _
+
+from piplmesh import panels
+from piplmesh.panels.horoscope import horoscope, models
```
Use relative imports.

domenc replied:
Like this? 
from . import horoscope, models

mitar replied:
Yes, you also need to add import from future. See how we do it elsewhere.

mitar commented on piplmesh/panels/horoscope/tasks.py:
```diff
@@ -0,0 +1,36 @@
+def update_horoscope():
+    update_all_horoscope()
+
+def update_all_horoscope():
```
Why defining this function if you are calling it only once?

domenc committed:
Some fix (imports, data filed, ..)

domenc commented:
DateFiled in manog DB, imports fix, some small fx and deletes.

mitar commented on piplmesh/panels/horoscope/horoscope.py:
```diff
@@ -0,0 +1,250 @@
+from __future__ import absolute_import
+
+from datetime import datetime
```
Import only modules! ;-)

mitar commented on piplmesh/panels/horoscope/horoscope.py:
```diff
@@ -0,0 +1,250 @@
+    """
+
+    languages = []
+    for object in get_all_horoscopes_providers():
```
Rename `object` to `horoscope`.

mitar commented on piplmesh/panels/horoscope/horoscope.py:
```diff
@@ -0,0 +1,250 @@
+    languages = []
+    for object in get_all_horoscopes_providers():
+        language = object.get_language()
+        assert (language not in languages)
```
No need for parenthesis. Write:

```
assert language not in languages, language
```

In this way error will also contain value `language`, so it will be easier to understand where exactly is a problem.

mitar commented on piplmesh/panels/horoscope/horoscope.py:
```diff
@@ -0,0 +1,250 @@
+
+class HoroscopeProviderBase(object):
+    """
+    Parent class for defined horoscope methods for getting horoscope in your language.
```
_Base class for horoscope providers._

mitar commented on piplmesh/panels/horoscope/horoscope.py:
```diff
@@ -0,0 +1,250 @@
+
+    def get_language(self):
+        """
+        Returns provider language.
```
_Returns provider's language._

mitar commented on piplmesh/panels/horoscope/horoscope.py:
```diff
@@ -0,0 +1,250 @@
+
+    def get_source_name(self):
+        """
+        Returns provider source name.
```
_Returns provider's source name._

mitar commented on piplmesh/panels/horoscope/horoscope.py:
```diff
@@ -0,0 +1,250 @@
+
+    def get_source_url(self, ):
+        """
+        Returns provider source url.
```
_Returns provider's source URL._

mitar commented on piplmesh/panels/horoscope/horoscope.py:
```diff
@@ -0,0 +1,250 @@
+
+class EnglishHoroscope(HoroscopeProviderBase):
+    """
+    Daily english horoscope from http://findyourfate.com
```
Dot at the end.

mitar commented on piplmesh/panels/horoscope/horoscope.py:
```diff
@@ -0,0 +1,250 @@
+    source_url = 'http://findyourfate.com/'
+
+    request_sign = {
+            'aries': 'Aries',
```
Indent one less.

mitar commented on piplmesh/panels/horoscope/horoscope.py:
```diff
@@ -0,0 +1,250 @@
+            'sagittarius': 'Sagittarius',
+            'capricorn': 'Capricorn',
+            'aquarius': 'Aquarius',
+        }
```
This `}` should be aligned with `request_sign`.

mitar commented on piplmesh/panels/horoscope/horoscope.py:
```diff
@@ -0,0 +1,250 @@
+
+class SlovenianHoroscope(HoroscopeProviderBase):
+    """
+    Daily slovenian horoscope from http://slovenskenovice.si
```
Do at the end.

mitar commented on piplmesh/panels/horoscope/horoscope.py:
```diff
@@ -0,0 +1,250 @@
+    source_url = 'http://www.slovenskenovice.si/'
+
+    request_sign = {
+            'aries': 'oven',
```
Indentation.

mitar commented on piplmesh/panels/horoscope/horoscope.py:
```diff
@@ -0,0 +1,250 @@
+    source_name = 'Slovenske novice'
+    source_url = 'http://www.slovenskenovice.si/'
+
+    request_sign = {
```
`provider_sign_names` would be probably enough?

mitar commented on piplmesh/panels/horoscope/horoscope.py:
```diff
@@ -0,0 +1,250 @@
+            'aquarius': 'vodnar',
+        }
+
+    months_names = {
```
`provider_month_names`

mitar commented on piplmesh/panels/horoscope/models.py:
```diff
@@ -0,0 +1,26 @@
+    ('aquarius', _("Aquarius")),
+)
+
+class Horoscope(mongoengine.Document):
```
Set fields to `required=True`.

mitar replied:
Some combination of fields should be `unique`, no? Like for each combination of `sign` and `language` should only one entry exist? Check [unique_with](http://mongoengine-odm.readthedocs.org/en/latest/guide/defining-documents.html#field-arguments).

mitar commented on piplmesh/panels/horoscope/panel.py:
```diff
@@ -0,0 +1,54 @@
+        horoscope_error = ''
+
+        if user.birthdate == None:
+            horoscope_error = _("Please, set your birthdate.")
```
I would not encode user feedback in the code. This is designers job to decide how to tell this to the user. You just set some constant and then check for it in the template. For example, here it would be much more user friendly to also provide a link to the page where user can set birthdate.

mitar commented on piplmesh/panels/horoscope/panel.py:
```diff
@@ -0,0 +1,54 @@
+        if user.birthdate == None:
+            horoscope_error = _("Please, set your birthdate.")
+        elif user.language not in horoscope.get_supported_languages():
+            horoscope_error = _("Horoscope is unavailable in your language.")
```
Same. We could invite users to tell us of horoscope source she enjoys and we could try to support it.

mitar replied:
(Or fork on GitHub and implement herself.)

mitar commented on piplmesh/panels/horoscope/panel.py:
```diff
@@ -0,0 +1,54 @@
+        horoscope_date = ''
+        horoscope_error = ''
+
+        if user.birthdate == None:
```
I still don't like this long if-structure. Why you don't simply update and return context immediately in those two special cases?

domenc replied:
Do you mean like that: 

if user.birthdate == None or user.language not in horoscope.get_supported_languages():
horoscope_error = True

or

if user.birthdate == None or user.language not in horoscope.get_supported_languages():
            return context # nothing to set

mitar replied:
```
if user.birthdate == None:
    context.update({
        'error_birthdate': True,
    })
    return context

if user.language not in horoscope.get_supported_languages(): # BTW, it would be better to use translation.get_language() here
    context.update({
        'error_language': True,
    })
    return context

...
```

mitar commented on piplmesh/panels/horoscope/tasks.py:
```diff
@@ -0,0 +1,34 @@
+    """
+
+    for horoscope_provider in horoscope.get_all_horoscopes_providers():
+        for sign in map(lambda s: s, horoscope.HOROSCOPE_SIGNS_DICT):
```
What is this map??? `lambda s: s` does nothing?

mitar commented on piplmesh/panels/horoscope/tasks.py:
```diff
@@ -0,0 +1,34 @@
+
+from celery import task
+
+from django.utils.encoding import smart_unicode
```
Import only modules!

mitar commented on piplmesh/panels/horoscope/models.py:
```diff
@@ -0,0 +1,26 @@
+class Horoscope(mongoengine.Document):
+    sign = mongoengine.StringField(choices=HOROSCOPE_SIGNS)
+    language = mongoengine.StringField(choices=settings.LANGUAGES)
+    description = mongoengine.StringField()
```
I really don't like `description` name here. Could you find some other name for this value?

domenc replied:
Text mybe?

mitar replied:
`forecast`?

mitar commented on piplmesh/panels/horoscope/tasks.py:
```diff
@@ -0,0 +1,34 @@
+    Update if exsits otherwise insert a new.
+    """
+
+    horoscope_language = smart_unicode(horoscope_language)
```
I think this should be done at the provider level.

domenc replied:
you mean smat_unicode or complet methos to move to horoscope.py??

mitar replied:
I mean that already `fetch_data` should assure that returned values are proper unicode. Not that caller have to do this.

domenc replied:
ok

mitar commented on piplmesh/panels/horoscope/horoscope.py:
```diff
@@ -0,0 +1,250 @@
+        languages.append(language)
+    return languages
+
+class HoroscopeProviderBase(object):
```
Define also `def fetch_data(self, sign):` method, which throws `NotImplemented` exception by default. So that API is fully documented.

mitar commented on piplmesh/panels/horoscope/tasks.py:
```diff
@@ -0,0 +1,34 @@
+        for sign in map(lambda s: s, horoscope.HOROSCOPE_SIGNS_DICT):
+            horoscope_language = horoscope_provider.get_language()
+
+            (horoscope_description, horoscope_sign, horoscope_date) = horoscope_provider.fetch_data(sign)
```
I really don't like this tuple thing. Can you at least return dict?

mitar commented on piplmesh/panels/horoscope/tasks.py:
```diff
@@ -0,0 +1,34 @@
+
+            insert_update_one_horoscope(horoscope_language, horoscope_description, horoscope_sign, horoscope_date)
+
+def insert_update_one_horoscope(horoscope_language, horoscope_description, horoscope_sign, horoscope_date):
```
No need to prefix with `horoscope_`. It is quite obvious.

domenc committed:
Some fix and update.

domenc commented:
Updated. For template HTML (fo displaying data and errors and for images) I create a new issue after this horoscope will be done.

domenc commented:
Mitar??

mitar commented on piplmesh/panels/horoscope/horoscope.py:
```diff
@@ -0,0 +1,264 @@
+from __future__ import absolute_import
+
+import datetime, re
+from urllib import urlopen
```
Import only modules.

mitar commented on piplmesh/panels/horoscope/horoscope.py:
```diff
@@ -0,0 +1,264 @@
+from urllib import urlopen
+
+from xml.dom import minidom
+from xml.dom.minidom import parseString
```
Import only modules.

mitar commented on piplmesh/panels/horoscope/horoscope.py:
```diff
@@ -0,0 +1,264 @@
+        return self.source_url
+
+    def fetch_data(self, sign):
+        return NotImplemented
```
`NotImplementedError`!

mitar commented on piplmesh/panels/horoscope/horoscope.py:
```diff
@@ -0,0 +1,264 @@
+        horoscope_xml = urlopen(horoscope_url).read()
+
+        date_string = parseString(horoscope_xml).getElementsByTagName('title')[1].toxml()
+        date_string = date_string.replace('<title>' + self.provider_sign_names[sign] + ' Horoscope for ', '').replace('</title>', '')
```
This is too error prone. You will have to capture date with regex.

mitar replied:
You could move date and forecast parsing to two methods. They get XML object (already) parsed and return values.

mitar commented on piplmesh/panels/horoscope/horoscope.py:
```diff
@@ -0,0 +1,264 @@
+        date_string = parseString(horoscope_xml).getElementsByTagName('title')[1].toxml()
+        date_string = date_string.replace('<title>' + self.provider_sign_names[sign] + ' Horoscope for ', '').replace('</title>', '')
+
+        date = datetime.datetime.strptime(date_string, '%A, %B %d, %Y')
```
This is not good, we learned that `%B` does not work reliably. Parsing based on locale does not really work. You should parse it manually, similar to how you do it for Slovene.

mitar commented on piplmesh/panels/horoscope/horoscope.py:
```diff
@@ -0,0 +1,264 @@
+
+    def fetch_data(self, sign):
+        horoscope_url = '%srss/dailyhoroscope-feed.asp?sign=%s' % (self.source_url, self.provider_sign_names[sign])
+        horoscope_xml = urlopen(horoscope_url).read()
```
Why don't you pare XML only once?

```
horoscope_xml = parseString(urlopen(horoscope_url).read())
```

Is there maybe even some function which reads from file descriptor? So you do not have to `read()` it first into the string for `parseString` to consume it?

mitar commented on piplmesh/panels/horoscope/horoscope.py:
```diff
@@ -0,0 +1,264 @@
+        date = datetime.datetime.strptime(date_string, '%A, %B %d, %Y')
+
+        forecast = parseString(horoscope_xml).getElementsByTagName('description')[1].toxml()
+        forecast = forecast.replace('<description>', '').replace('</description>', '')
```
No, get content of `description`. Without tags around. There is probably some function for this?

mitar commented on piplmesh/panels/horoscope/horoscope.py:
```diff
@@ -0,0 +1,264 @@
+        index_line_date = horoscope_xml.find(search_start_date)
+        index_start_date = index_line_date+len(search_start_date)
+        index_end_date = horoscope_xml.find(search_end_date)
+        date_string_html = horoscope_xml[index_start_date:index_end_date]
```
Move date parsing to a method.

mitar commented on piplmesh/panels/horoscope/horoscope.py:
```diff
@@ -0,0 +1,264 @@
+            'date': date,
+            'forecast': encoding.smart_unicode(forecast),
+            'sign': encoding.smart_unicode(sign),
+            'language': encoding.smart_unicode(self.language),
```
But caller already know this? Why it is again returned?

mitar commented on piplmesh/panels/horoscope/horoscope.py:
```diff
@@ -0,0 +1,264 @@
+        return {
+            'date': date,
+            'forecast': encoding.smart_unicode(forecast),
+            'sign': encoding.smart_unicode(sign),
```
But caller already know this? Why it is again returned?

mitar commented on piplmesh/panels/horoscope/panel.py:
```diff
@@ -0,0 +1,66 @@
+from django.utils.translation import ugettext_lazy as _
+
+from piplmesh import panels
+from . import horoscope, models
```
One empty line before this.

mitar commented on piplmesh/panels/horoscope/panel.py:
```diff
@@ -0,0 +1,66 @@
+
+        if translation.get_language() not in horoscope.get_supported_languages():
+            context.update({
+                'header': _("Today's horoscope"),
```
You can move setting header to the beginning of this method. Because it is shared with all code paths.

mitar commented on piplmesh/panels/horoscope/panel.py:
```diff
@@ -0,0 +1,66 @@
+        horoscope_date = ''
+        horoscope_date_string = ''
+
+        if user.birthdate == None:
```
`if not user.birthdate`

mitar replied:
If you want to compare with `None`, always do `if user.birthdate is None`.

mitar commented on piplmesh/panels/horoscope/panel.py:
```diff
@@ -0,0 +1,66 @@
+
+        user_sign = horoscope.get_horoscope_sign(user.birthdate.day, user.birthdate.month)
+
+        horoscope_object = models.Horoscope.objects(
```
Why `object` suffix?

mitar commented on piplmesh/panels/horoscope/horoscope.py:
```diff
@@ -0,0 +1,264 @@
```
I would rename this file to `providers.py`.

mitar commented on piplmesh/panels/horoscope/horoscope.py:
```diff
@@ -0,0 +1,264 @@
+
+    return HOROSCOPE_PROVIDERS
+
+def get_horoscope_provider(language):
```
Probably `horoscope` word is unnecessary here?

`piplmesh.panels.horoscope.providers.get_provider` tells enough that it is provider for horoscope.

mitar commented on piplmesh/panels/horoscope/horoscope.py:
```diff
@@ -0,0 +1,264 @@
+        else:
+            return 'aquarius'
+
+def get_all_horoscopes_providers():
```
Probably `horoscope` word is unnecessary here?

mitar commented on piplmesh/panels/horoscope/panel.py:
```diff
@@ -0,0 +1,66 @@
+        horoscope_sign = horoscope.HOROSCOPE_SIGNS_DICT[user_sign]
+        horoscope_date = horoscope_object[0].date
+
+        horoscope_date_string = "%s.%s.%s" % (horoscope_date.day, horoscope_date.month, horoscope_date.year)
```
??? You should render this in template!

mitar commented on piplmesh/panels/horoscope/panel.py:
```diff
@@ -0,0 +1,66 @@
+                'error_data': True,
+            })
+            return context
+
```
You should check if horoscope data is too old (more than one day?).

domenc committed:
Fixed. Regex fro XML in English horoscope (in Sloven its to complex, because am parsing HTML with CSS...), add horoscope date for checing if is older for 2 days, ...

domenc commented:
All fixed. Add checking horoscope date (if is 2 days out -> too old to display),.. regex for english horoscope (for slove is to complex because is html and not xml),..

domenc commented:
Mitar, why are you activity in other repositories and don't here any more?

mitar commented:
Sorry. Busy here. We got permission to use Slovenske novice horoscope for noncommercial academic use. :-)

Have you checked terms of use for the other (English) source?

domenc commented:
Yes, I write to them and have positive feedback to use their horoscope

mitar commented:
Can you please forward their e-mail to me?

domenc commented:
I sent message to your email.

mitar commented on piplmesh/panels/horoscope/panel.py:
```diff
@@ -0,0 +1,74 @@
+
+from piplmesh import panels
+
+from . import providers, models
```
Alphabetical order.

mitar commented on piplmesh/panels/horoscope/panel.py:
```diff
@@ -0,0 +1,74 @@
+            })
+            return context
+
+        if datetime.datetime.now() > (horoscope[0].date + datetime.timedelta(days=2)):
```
2 should be a top level module constant.

mitar commented on piplmesh/panels/horoscope/panel.py:
```diff
@@ -0,0 +1,74 @@
+
+        if datetime.datetime.now() > (horoscope[0].date + datetime.timedelta(days=2)):
+            context.update({
+                'error_date': True,
```
I would rename this to `error_obsolete` because `error_date` is too close to `error_data`.

mitar commented on piplmesh/panels/horoscope/panel.py:
```diff
@@ -0,0 +1,74 @@
+
+        user_sign = providers.get_horoscope_sign(user.birthdate.day, user.birthdate.month)
+
+        horoscope = models.Horoscope.objects(
```
ojects.get()? And use exception if missing?

mitar commented on piplmesh/panels/horoscope/panel.py:
```diff
@@ -0,0 +1,74 @@
+            })
+            return context
+
+        horoscope_forecast = horoscope[0].forecast
```
Because `horoscope[0]` is ugly.

mitar replied:
There is no need to define temporary variables. You could just populate context with them.

mitar commented on piplmesh/panels/horoscope/panel.py:
```diff
@@ -0,0 +1,74 @@
+    def get_context(self, context):
+        user = context['user']
+
+        horoscope_forecast = ''
```
I don't see why those temporary variables are necessary?

mitar commented on piplmesh/panels/horoscope/panel.py:
```diff
@@ -0,0 +1,74 @@
+
+        horoscope = models.Horoscope.objects(
+            sign = user_sign,
+            language = user.language
```
No, not `user.language` (although it is the same as user.language, but this is just by our design and middleware), but `translation.get_language()`. So simply you display in current language. We might someday allow to override language without changing the preference. Or something.

mitar commented on piplmesh/panels/horoscope/providers.py:
```diff
@@ -0,0 +1,279 @@
+
+from . import models
+
+HOROSCOPE_SIGNS_DICT = dict(models.HOROSCOPE_SIGNS)
```
I would move this to `models.py`, so everything is at the same place.

mitar commented on piplmesh/panels/horoscope/panel.py:
```diff
@@ -0,0 +1,74 @@
+        horoscope_date = horoscope[0].date
+
+        context.update({
+            'horoscope_forecast': horoscope_forecast,
```
So just move above lines here, to update. So instead of:

```
'horoscope_forecast': horoscope_forecast,
```

simply:

```
'horoscope_forecast': horoscope.forecast
```

and so on. (There is no `[0]` if you will use `get()`.)

mitar commented on piplmesh/panels/horoscope/panel.py:
```diff
@@ -0,0 +1,74 @@
+            })
+            return context
+
+        if translation.get_language() not in providers.get_supported_languages():
```
You could just do:

```
provider = providers.get_provider(translation.get_language())
```

and catch the exception. Or if found, use `provider` later on in context update. Instead of multiple calls.

mitar commented on piplmesh/panels/horoscope/providers.py:
```diff
@@ -0,0 +1,279 @@
+
+        return self.source_name
+
+    def get_source_url(self, ):
```
Why `,` here?

mitar commented on piplmesh/panels/horoscope/providers.py:
```diff
@@ -0,0 +1,279 @@
+            'forecast': encoding.smart_unicode(self.forecast_parsing(horoscope_xml)),
+        }
+
+    def date_parsing(self, xml, sign):
```
This is not good. Please just make one regex to match date out of the text, don't do it like you are doing it now. Removing parts. This crazy. Can you provide me with an example string and I can help you make regex.

domenc replied:
Yes, please help me with parsing - example of XML: http://findyourfate.com/rss/dailyhoroscope-feed.asp?sign=Aries
Tnx

mitar replied:
`tree.findtext('.//item/title')` returns title text, which is for example `Aries Horoscope for Wednesday, August 01, 2012`.

You want to get out `August`, `01`, and `2012`. So this returns you that:

```
import re
date_pattern = re.compile(r'(\w+)\s+(\d+),\s+(\d+)$')
date_pattern.search('Aries Horoscope for Wednesday, August 01, 2012').groups()
```

mitar replied:
Please read the regex there and explain it to me. Not just use it. Try to understand it.

mitar commented on piplmesh/panels/horoscope/providers.py:
```diff
@@ -0,0 +1,279 @@
+        date_string_month = re.sub('[0-9, ]', '', re.sub(r'\w{5,10}, ', '', date_string_text))
+        date_string_year = re.findall(r'\d{4}', date_string_text)[0]
+        date_string = '%s. %s %s' % (date_string_day, self.provider_month_names[date_string_month], date_string_year)
+        date = datetime.datetime.strptime(date_string, '%d. %B %Y')
```
Ehm, you construct string from parsed values just to parse it again?

I think `datetime.datetime.date(year, month, day)` would be enough. Just get year, month, and day values.

domenc replied:
Not good, because month are string - name of month and not a number

mitar replied:
So what? You are mapping above from string to string in `provider_month_names`. Why not map from string to number instead?

mitar commented on piplmesh/panels/horoscope/providers.py:
```diff
@@ -0,0 +1,279 @@
+
+    def fetch_data(self, sign):
+        horoscope_url = '%slifestyle/astro/%s' % (self.source_url, self.provider_sign_names[sign])
+        horoscope_xml = urllib.urlopen(horoscope_url).read()
```
Decide how you want variables to be named. Here you have `_xml` suffix, but it is not really a XML object. Maybe something else would be better.

mitar replied:
You cannot parse this page? And extract in similar way? Maybe not parse as XML, but use some HTML parser. So that you get DOM out of it. And then just navigate to those elements you want.

domenc replied:
Any better idea for this: http://slovenskenovice.si/lifestyle/astro/oven??

mitar commented on piplmesh/panels/horoscope/tasks.py:
```diff
@@ -0,0 +1,26 @@
+
+from celery import task
+
+from . import providers, models
```
Alphabetical order.

mitar commented on piplmesh/panels/horoscope/templates/panel/horoscope/panel.html:
```diff
@@ -0,0 +1,27 @@
+{% extends "panel/panel.html" %}
+
+{% load i18n static sekizai_tags %}
```
Please use `staticfiles` instead of `static`. Crazy Django thing.

mitar commented on piplmesh/panels/horoscope/templates/panel/horoscope/panel.html:
```diff
@@ -0,0 +1,27 @@
+
+{% block content %}
+    {% addtoblock "css" %}<link href="{% static "piplmesh/panel/horoscope/panel.css" %}" rel="stylesheet" type="text/css" media="screen" />{% endaddtoblock %}
+    {% if user.is_authenticated %}
```
Why is this necessary?

domenc replied:
If user not authenticated then we display message for login

mitar replied:
Why? Why not simply display horoscope, if user has birthdate, or error if not. Why is important if it is authenticated or not? Code checks already if birthdate is set and this is all we are interested in.

domenc replied:
If I'm not mistaken it also has an unregistered user configured birthday to 1.1.1970

mitar replied:
It shouldn't be. It should be None, undefined.

mitar commented on piplmesh/panels/horoscope/templates/panel/horoscope/panel.html:
```diff
@@ -0,0 +1,27 @@
+            {% else %}
+                <span class="horoscope_info">{{ horoscope_sign }}</span><br />
+                <span class="horoscope_forecast">{{ horoscope_forecast }}</span><br />
+                <span class="horoscope_date">{% trans "Date" %}: {{ horoscope_date.day }}.{{ horoscope_date.month }}.{{ horoscope_date.year }}</span><br />
```
No, don't format it like this. Date formatting should be done so that it depends on the language used. Find one Django template filter for that. :-)

mitar commented on piplmesh/panels/horoscope/templates/panel/horoscope/panel.html:
```diff
@@ -0,0 +1,27 @@
+                <span class="horoscope_info">{{ horoscope_sign }}</span><br />
+                <span class="horoscope_forecast">{{ horoscope_forecast }}</span><br />
+                <span class="horoscope_date">{% trans "Date" %}: {{ horoscope_date.day }}.{{ horoscope_date.month }}.{{ horoscope_date.year }}</span><br />
+                <span class="horoscope_url">{% trans "Source" %}: <a href="{{ horoscope_source_url }}">{{ horoscope_source_name }}</a></span>
```
Use transblock tag and pass url and name as parameters. You cannot concatenate strings like this because maybe in some language it should be displayed differently. Sam for date above.

mitar commented on piplmesh/panels/horoscope/tasks.py:
```diff
@@ -0,0 +1,26 @@
+
+def insert_update_one_horoscope(horoscope, sign, language):
+    """
+    Update if exsits otherwise insert a new.
```
Typo in _exsits_.

mitar commented on piplmesh/panels/horoscope/tasks.py:
```diff
@@ -0,0 +1,26 @@
+
+            insert_update_one_horoscope(horoscope_data, sign, horoscope_provider.get_language())
+
+def insert_update_one_horoscope(horoscope, sign, language):
```
Can you order function parameters in the same order as they are define in the document?

mitar replied:
Do we really need a separate function?

mitar commented on piplmesh/panels/horoscope/models.py:
```diff
@@ -0,0 +1,26 @@
+from django.conf import settings
+from django.db import models
```
Django models? This cannot be right.

mitar commented on piplmesh/panels/horoscope/tasks.py:
```diff
@@ -0,0 +1,26 @@
+    """
+
+    # Try update, if failed insert a new object
+    if not models.Horoscope.objects(language=language, sign=sign).update(set__forecast=horoscope['forecast'], set__date=horoscope['date']):
```
Does not really work. So there is still a race condition. What you want to do here is called upsert (insert or update) and MongoEngine supports this natively. This would do the trick:

```
models.Horoscope.objects(language=language, sign=sign).update(set__forecast=horoscope['forecast'], set__date=horoscope['date'], upsert=True)
```

And of course this line does not need its own function.

mitar commented:
Otherwise, great! :-) All those are now minor comments.

domenc committed:
Update some minor code

domenc commented:
Update some minor code...
TODO with Mitar: better regex for parsing

mitar commented on piplmesh/panels/horoscope/panel.py:
```diff
@@ -0,0 +1,71 @@
+
+        try:
+            provider = providers.get_provider(translation.get_language())
+        except:
```
Don/t catch all exceptions!

domenc replied:
OK, only KeyError from get_provider()??

mitar replied:
Will not answer this question.

mitar commented on piplmesh/panels/horoscope/panel.py:
```diff
@@ -0,0 +1,71 @@
+
+        try:
+            horoscope = models.Horoscope.objects.get(
+                sign = user_sign,
```
No spaces around `=` in call arguments.

mitar commented on piplmesh/panels/horoscope/panel.py:
```diff
@@ -0,0 +1,71 @@
+        try:
+            horoscope = models.Horoscope.objects.get(
+                sign = user_sign,
+                language = translation.get_language()
```
Comma at the end.

mitar commented on piplmesh/panels/horoscope/panel.py:
```diff
@@ -0,0 +1,71 @@
+                sign = user_sign,
+                language = translation.get_language()
+            )
+        except DoesNotExist:
```
Where is `DoesNotExist` defined?

domenc replied:
At mongoengine.. http://mongoengine-odm.readthedocs.org/en/latest/apireference.html?highlight=exception#mongoengine.queryset.QuerySet.get

mitar replied:
Ehm. OK, to ask differently: where it is imported?

mitar commented on piplmesh/panels/horoscope/panel.py:
```diff
@@ -0,0 +1,71 @@
+            })
+            return context
+
+        if not horoscope:
```
Why would there be no horoscope at this point?

domenc replied:
I forgot, before it was necessary, because i havent except Does not exits

mitar commented on piplmesh/panels/horoscope/panel.py:
```diff
@@ -0,0 +1,71 @@
+            })
+            return context
+
+        if datetime.datetime.now() > (horoscope.date + datetime.timedelta(days=HOROSCOPE_DATE_TOO_OLD)):
```
No need for parenthesis around the sum.

mitar commented on piplmesh/panels/horoscope/providers.py:
```diff
@@ -0,0 +1,277 @@
+        return date
+
+    def forecast_parsing(self, xml):
+        return xml.getElementsByTagName('description')[1].firstChild.data
```
So this can be done with:

```
from lxml import etree
tree = etree.parse('http://findyourfate.com/rss/dailyhoroscope-feed.asp?sign=Aries')
return tree.findtext('.//item/description')
```

mitar replied:
(You do not really need an additional function for this.)

mitar commented:
Please create also a manage.py command which triggers horoscope fetching immediately (creates a task and waits for it to finish).

mitar commented on piplmesh/panels/horoscope/providers.py:
```diff
@@ -0,0 +1,277 @@
+        'december': 'December',
+    }
+
+    def fetch_data(self, sign):
```
```
from lxml import html
tree = html.parse('http://slovenskenovice.si/lifestyle/astro/oven')
print tree.findtext('.//div[@id="horoscope-sign-right"]//div[@class="view-content"]//span')
print tree.findtext('.//div[@id="horoscope-sign-right"]//div[@class="view-content"]//strong')
```

You should really learn XPath. So please explain XPath above.

And I will leave regex for parsing date out of the string to you.

mitar replied:
Please explain XPaths above.

domenc replied:
Find DIV element with id "horoscope-sign-right" and inside this DIV element search for another DIV with class "view-contetn".
First XPath display SPAN tag, second display STRONG tag.

mitar commented on piplmesh/panels/horoscope/panel.py:
```diff
@@ -0,0 +1,71 @@
+
+class HoroscopePanel(panels.BasePanel):
+    def get_context(self, context):
+        user = context['user']
```
Please use `self.request.user` here.

mitar commented on piplmesh/panels/horoscope/tasks.py:
```diff
@@ -0,0 +1,18 @@
+        for sign in models.HOROSCOPE_SIGNS_DICT:
+            horoscope_data = horoscope_provider.fetch_data(sign)
+
+            models.Horoscope.objects(language=horoscope_provider.get_language(), sign=sign).update(set__forecast=horoscope_data['forecast'], set__date=horoscope_data['date'], upsert=True)
```
I was thinking. It would be interesting to store all horoscope data, also for the past. I am interesting to see how much they change through time or if our provides are re-using them and so on. So override only horoscope of the same day. You know how to change above query to store horoscope every day?

domenc replied:
Maybe someday :D - this could be a new issue :D....

First, remove unique_with='sign'
Then objects(date=current_date).update(....)...

OK?

mitar replied:
Why remove `unique_with`? You should extend it to include date, too.

I see that you don't understand how above query works. Please take some time to learn about that and to understand it.

mitar commented on piplmesh/panels/horoscope/panel.py:
```diff
@@ -0,0 +1,71 @@
+        user_sign = providers.get_horoscope_sign(user.birthdate.day, user.birthdate.month)
+
+        try:
+            horoscope = models.Horoscope.objects.get(
```
You will have to order those objects by date and get only `first()`, if we will store also past horoscopes.

domenc committed:
Fixs

domenc commented:
Some update.

And please, not making you go along the new requirements. 

The basic requirement was:
"""
Implement horoscope panel on home page based on user's birthday. Find some public domain daily horoscope in English and Slovene, or if not public, Creative Commons, of if not commons, something you use. And at the bottom of the horoscope panel put attribution link to the source.
"""
Therefore, manage.py and so on, leaving for a new issue tickets.
Tnx.

mitar commented:
As you want. But then don't complain there are not enough ideas to work on.

mitar commented:
Please open the relevant tickets for things you will not do. So that others can see what is possible to do.

mitar commented:
(This pull request cannot be automatically merged, please update from the main repository.)

mitar commented:
It looks good now.

mitar commented on piplmesh/panels/horoscope/models.py:
```diff
@@ -0,0 +1,27 @@
+
+HOROSCOPE_SIGNS_DICT = dict(HOROSCOPE_SIGNS)
+
+class Horoscope(mongoengine.Document):
```
Please order fields in: date, language, sign, forecast. In this way they are ordered from less specific to more, and then at the end the content.

Also add indexes to date, language and sign fields.

Order those fields also elsewhere in the code like this (for example, in arguments passing to functions).

mitar commented on piplmesh/panels/horoscope/panel.py:
```diff
@@ -0,0 +1,65 @@
+            })
+            return context
+
+        if datetime.datetime.now() > horoscope.date + datetime.timedelta(days=HOROSCOPE_DATE_TOO_OLD):
```
`HOROSCOPE_DATE_TOO_OLD` -> `HOROSCOPE_OBSOLETE` if you are using `error_obsolete` bellow. Or change to `error_date_too_old` bellow.

mitar commented on piplmesh/panels/horoscope/providers.py:
```diff
@@ -0,0 +1,247 @@
+from __future__ import absolute_import
+
+import datetime, re, urllib
```
Is `urllib` needed?

mitar commented on piplmesh/panels/horoscope/providers.py:
```diff
@@ -0,0 +1,247 @@
+        horoscope_html_tree = html.parse(horoscope_url)
+
+        date_html_string = tree.findtext('.//div[@id="horoscope-sign-right"]//div[@class="view-content"]//span')
+        date_string = re.compile(r'(\w+)\s+(\w+)\s+(\d+).\s+(\w+):\s+$')
```
Ehm. Missing some more code here?

mitar replied:
Wouldn't it be better to name first variable `date_string` and the other `date_parsed` or something?

mitar replied:
You don't need first two capture groups here in reges. `r'(\d+).\s+(\w+):\s+$'` should be probably enough. :-)

mitar commented on piplmesh/panels/horoscope/tasks.py:
```diff
@@ -0,0 +1,18 @@
+        for sign in models.HOROSCOPE_SIGNS_DICT:
+            horoscope_data = horoscope_provider.fetch_data(sign)
+
+            models.Horoscope.objects(language=horoscope_provider.get_language(), sign=sign, date=horoscope_data['date']).update(set__forecast=horoscope_data['forecast'], upsert=True)
```
Good! But order fields!

mitar commented on piplmesh/panels/horoscope/tasks.py:
```diff
@@ -0,0 +1,18 @@
+def update_horoscope():
+    """
+    Function for updating all languages avaiable horoscope.
+    Update if exists otherwise insert a new.
```
_Update_ -> _Updates_, _insert_ -> _inserts_.

mitar commented on piplmesh/panels/horoscope/templates/panel/horoscope/panel.html:
```diff
@@ -0,0 +1,27 @@
+        {% else %}
+            <span class="horoscope_info">{{ horoscope_sign }}</span><br />
+            <span class="horoscope_forecast">{{ horoscope_forecast }}</span><br />
+            {% blocktrans with date=horoscope_date|date:"DATE_FORMAT" title="Date"|title %}
```
No need for "title". This you can leave simply in the blocktrans content.

Also don't have newlines in blocktrans. It is really hard to translate. Put <span> directly inside of blocktrans. Yes, it is ugly, but ...

mitar commented:
So it looks good now. Clean up the things and test it. And this is it.

mitar commented:
This pull request cannot be automatically merged. Please merge.

Will you finish the rest?

domenc commented:
yes, tomorrow. Current i am at Bovec.

## 

Domen Čebulj
http://cebulj.net
http://olsevek.si

domenc committed:
Add manage.py command for manual udpoating horoscope and some small fixes

domenc committed:
requirements

domenc committed:
Add requirements

domenc commented:
Add manage.py command and other fixes.

mitar commented:
I still cannot merge it. Have you merged with latest from the main repository?

mitar commented on piplmesh/panels/horoscope/providers.py:
```diff
@@ -0,0 +1,249 @@
+        horoscope_url = '%srss/dailyhoroscope-feed.asp?sign=%s' % (self.source_url, self.provider_sign_names[sign])
+        horoscope_tree = etree.parse(horoscope_url)
+
+        date_string = re.compile(PATTERN_DATE_EN).search(horoscope_tree.findtext('.//item/title'))
```
Nooo. You compile the pattern only once. And store compiled pattern in the top-level constant variable. :-) Compiling the pattern is the process we would not like to repeat again and again.

mitar commented:
OK. It looks good otherwise. There are some minor things still, but I will not bother you with that anymore. Please do a merge and I will pull it in.

mitar commented on piplmesh/panels/horoscope/management/commands/updatehoroscope.py:
```diff
@@ -0,0 +1,15 @@
+        Manually update horoscopes with manage.py command.
+        """
+
+        self.stdout.write('Start the update manually horoscopes. Please wait ...')
```
(You should write only when verbosity flag was set.)

mitar commented on piplmesh/panels/horoscope/management/commands/updatehoroscope.py:
```diff
@@ -0,0 +1,15 @@
+
+        self.stdout.write('Start the update manually horoscopes. Please wait ...')
+        self.stdout.write('\n\r')
+        tasks.update_horoscope()
```
(You should create an asynchronous task and wait for it. So that it is done through Celery. Not directly.)

mitar commented on piplmesh/panels/horoscope/templates/panel/horoscope/panel.html:
```diff
@@ -0,0 +1,28 @@
+        {% else %}
+            <span class="horoscope_info">{{ horoscope_sign }}</span><br />
+            <span class="horoscope_forecast">{{ horoscope_forecast }}</span><br />
+            {% blocktrans with date=horoscope_date|date:"DATE_FORMAT" title="Data" %}
```
"Data"? :-)

Please move this out of the blocktrans arguments.

mitar commented:
OK. The merge is simple. And it looks good. Great job.

domenc commented:
I think so..

(Yesterday I run command:
    - git checkout master
    - git pull
    - git checkout issue209_2 
    - git rebase master
)

I think that only file which you can't merge is requirements.txt (I think that I broke it :( )

TODO this afternoon:
    - save compile object
    - verbosity flag (only if set then print a message)
    - asynchronous task (call direcly calery for update horoscope??)
    - "Data" -> "Date", but if I remove from blocktrans arguments and add only that {{ trans "Date" }} is not work

I do not know how to fix mergeing this - requirements.txt (requirements.txt and settings.py was edit (in each file only add one line, all other files is new))

mitar commented:
Hm. Why rebase? Merge? `git merge master` Don't do rebase.

Yes, I cannot merge requirements. You have to do it for me.

> asynchronous task

You call delay and then get to wait for it to finish. Read celery tutorial to see what I mean.

You do not have to do `{{ trans }}` inside of blocktrans - whole block is already marked for translation. This is the main point. You just write any long text you want translated there.

domenc commented:
But if i run git merge master under my brench I broke my fork for other users??

OK

mitar commented:
No? You just merge master into your branch. So you don't merge INTO master. But FROM master into your branch. So don't get your master branch dirty! But you have to get into your branch content from master.

domenc commented:
OK, I will try this afternoon when i come home.
Tnx

domenc committed:
Merge branch 'master' into issue209_
qonflicts:
	requirements.txt

domenc commented:
New branch for this issue - #263

I broke branch :(

mitar commented:
How broke? Otherwise, until you push, you can undo and fix everything.

domenc commented:
I think I drty my branch.. So I open a new branch with clenly code...
